test/support/metrics_case.ex

```diff
@@ -0,0 +1,118 @@
+defmodule RequestCache.Support.MetricsCase do
+  @moduledoc """
+  This module defines the setup for tests requiring metrics tests.
+
+  Available tags:
+    - `telemetry_events`: list the Telemetry events to listen
+    - `metrics`: Specify the list of Telemetry.Metrics to used (format: [Module, :function, [args]])
+
+  Available assertions:
+    - `assert_telemetry_event(name, measurements, metadata \\ %{})`
+    - `assert_metric(name, measurement, metadata \\ %{})`
+
+  ## Example:
+
+  @tag telemetry_events: [[:user, :subscription, :email_confirmation]],
+       metrics: [RequestCache.Support.Metrics, :metrics, []]
+  test "my test" do
+    ...
+    assert_telemetry_event([:user, :subscription, :email_confirmation], %{count: 1}, %{result: :error})
+    assert_metric([:user, :subscription, :email_confirmation, :count], 1, %{success: false})
+  end
+
+  Sauce: https://elixirforum.com/t/testing-and-telemetry-events-how-to-test-if-they-are-sent/28273/4
+  """
+  use ExUnit.CaseTemplate
+
+  using do
+    quote do
+      import RequestCache.Support.MetricsCase
+    end
+  end
+
+  setup tags do
+    if telemetry_events = tags[:telemetry_events] do
+      metrics = get_metrics_from_tag(tags)
+
+      handler_config = %{}
+      |> Map.put(:metrics, metrics)
+      |> Map.put(:self, self())
+      |> Map.put(:groups, Enum.group_by(metrics, & &1.event_name))
+
+      :telemetry.attach_many(
+        tags[:test],
+        telemetry_events,
+        &__MODULE__.handle_event/4,
+        handler_config
+      )
+    end
+
+    :ok
+  end
+
+  def handle_event(name, measurements, metadata, %{groups: groups, self: self} = _config) do
```

Comment:
I think you should encapsulate this in a module instead of a cache case, this feels like it's doing a bit too much magic that makes the tests hard to understand how to write and work with, whereas if it was just a module under support, you would be able to call it and just see how things work without cross referencing files 🤔  what are your thoughts?

Reply:
Wanna settle on two small helper methods? ;-)